chainerx_cc/chainerx/routines/linalg.cc

```diff
@@ -189,4 +192,161 @@ Array Inverse(const Array& a) {
+        } else {
+            u_shape = Shape{m, mn};
+            vt_shape = Shape{mn, n};
+        }
+    } else {
+        u_shape = Shape{0};
+        vt_shape = Shape{0};
+    }
+
+    u = Empty(u_shape, a.dtype(), a.device());
+    vt = Empty(vt_shape, a.dtype(), a.device());
+    s = Empty(Shape{mn}, a.dtype(), a.device());
+
+    {
+        NoBackpropModeScope scope{};
+        a.device().backend().CallKernel<SvdKernel>(a, u, s, vt, full_matrices);
+    }
+
+    // Reference:
+    // https://j-towns.github.io/papers/svd-derivative.pdf
+    {
+        BackwardBuilder bb{"svd", a, {u, s, vt}};
+        if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+            bt.Define([a_tok = bb.RetainInput(0),
+                       u_tok = bb.RetainOutput(0),
+                       s_tok = bb.RetainOutput(1),
+                       vt_tok = bb.RetainOutput(2),
+                       full_matrices,
+                       compute_uv](BackwardContext& bctx) {
+                if (full_matrices) {
+                    throw ChainerxError{"ChainerX SVD differentiation is not implemented for full_matrices=true."};
+                }
+                if (!compute_uv) {
+                    throw ChainerxError{"ChainerX SVD differentiation cannot be computed with compute_uv=false."};
+                }
+
+                const Array& a = bctx.GetRetainedInput(a_tok);
+                const Array& u = bctx.GetRetainedOutput(u_tok);
+                const Array& s = bctx.GetRetainedOutput(s_tok);
+                const Array& vt = bctx.GetRetainedOutput(vt_tok);
+
+                auto m = a.shape()[0];
+                auto n = a.shape()[1];
+                auto k = s.shape()[0];
+
+                const Array& gu = bctx.output_grad(0).has_value() ? *bctx.output_grad(0) : Zeros(u.shape(), a.dtype(), a.device());
+                const Array& gsigma = bctx.output_grad(1).has_value() ? *bctx.output_grad(1) : Zeros(s.shape(), a.dtype(), a.device());
+                const Array& gvt = bctx.output_grad(2).has_value() ? *bctx.output_grad(2) : Zeros(vt.shape(), a.dtype(), a.device());
+
+                Array v = vt.Transpose();
+                Array gv = gvt.Transpose();
+
+                Array sigma_term = Dot(Dot(u, Diag(gsigma)), vt);
+
+                Array ut = u.Transpose();
+                Array im = Eye(m, m, 0, a.dtype(), a.device());
+                Array in = Eye(n, n, 0, a.dtype(), a.device());
+                Array sigma_mat = Diag(s);
+                Array sigma_mat_inv = Diag(Reciprocal(s));
+                Array sigma_sq = Power(s, 2);
+                Array F = ExpandDims(sigma_sq, 0) - ExpandDims(sigma_sq, 1);
+                // Invert values of F, and fill the diagonal with 0s.
+                // F has 0s on the diagonal, therefore fill it first with infinity.
+                Array mask = Eye(F.shape()[0], F.shape()[1], 0, Dtype::kBool, a.device());
+                F = Where(mask, INFINITY, F);
```

Comment:
Probably safer to use constants from `numeric_limits.h` for these types of values.